### ISSUE TITLE
`with` style context manager for Parsl DFK

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#3258]
**Is your feature request related to a problem? Please describe.**

Users should be able to use parsl DFKs using the `with` statement, to cause the DFK to be cleaned up when it is finished with.

Previously, if a user did not explicitly clean up a DFK, an `atexit` handler would do this when the process exits. See issue #3104 for why that doesn't work in recent Python, PR #3165 for the PR that removes that handling and issue #3251 for an example of a user encountering this as a practical problem.

I would like some syntax like this:

```
with parsl.load(config):
    f = my_app()
    f.result()
# implicit contextmanager cleanup of the loaded DFK happens here by calling dfk.cleanup()
```

In addition to introducing this context manager, documentation should be changed to encourage users to use this syntax instead of explicitly calling cleanup() in most scripted cases (although not always, for example, I think syntactically this doesn't work with jupyter notebooks as I think a `with` block cannot span multiple notebook cells)

**Describe alternatives you've considered**
It might be possible to rework exit handling in an atexit handler to still be able to do shutdown, but I don't feel like that will be easy - because a lot of stuff can happen at DFK shutdown.

**Additional context**
This is coming from changed (more strict) behaviour around threads and processes in Python 3.12.

**Comment by Pazil24** [comment]
@benclifford I'd like to work on this issue. Thank you.

**Comment by marameref** [comment]
@benclifford I will work on this issue if thats ok

**Comment by benclifford** [comment]
ok if multiple people are working on this, you might all like to comment here on your approaches and try to work together a bit

**Comment by marameref** [comment]
> ok if multiple people are working on this, you might all like to comment here on your approaches and try to work together a bit

ok @benclifford, i just cloned the repo and trying to better understand the issue, will give details of my approach asap

**Comment by Pazil24** [comment]
I am still getting a hang of Parsl from the tutorials. 
To work on this project to add the DFK context management, I :
1. Forked the Parsl repo
2. Cloned the repo locally using git clone <forked repo url>
3. Then I added the remote branch named upstream and the remote url
4. I merged it with master to get all updates

I am currently down setting up my dev environment and created a new working branch to start adding changes.
